vllm/model_executor/layers/quantization/aqlm.py

```diff
@@ -0,0 +1,236 @@
+                "tensor parallel size.")
+        if output_size_per_partition % self.quant_config.out_group_size != 0:
+            raise ValueError(
+                "The output size is not aligned with the quantized "
+                "weight shape. This can be caused by too large "
+                "tensor parallel size.")
+
+        codes = Parameter(
+            torch.empty(
+                # There could actually be two pack factors, one along input and one along output,
+                # but we don't currently support out_group_size,
+                # and only the one along output needs to be marked with "packed_dim".
+                # in order for QKVLinear to work.
+                output_size_per_partition,
+                input_size_per_partition // self.quant_config.pack_factor,
+                self.quant_config.num_codebooks,
+                dtype=get_int_dtype(self.quant_config.nbits_per_codebook),
+            ),
+            requires_grad=False,
+        )
+
+        set_weight_attrs(
+            codes,
+            {
+                "input_dim": 1,
+                "output_dim": 0,
+                "packed_dim": 1,
+                "pack_factor": self.quant_config.pack_factor,
+                "output_partition_sizes": output_partition_sizes
+            },
+        )
+
+        codebooks = Parameter(
+            torch.empty(
+                self.quant_config.num_codebooks * len(output_partition_sizes),
+                2**self.quant_config.nbits_per_codebook,
+                self.quant_config.out_group_size,
+                self.quant_config.in_group_size,
+                dtype=params_dtype,
+            ),
+            requires_grad=False,
+        )
+        set_weight_attrs(
+            codebooks,
+            {
+                "shard_dim": 0,
+            },
+        )
+
+        scales = Parameter(
+            torch.empty(
+                (
+                    output_size_per_partition //
+                    self.quant_config.out_group_size,
+                    1,
+                    1,
+                    1,
+                ),
+                dtype=params_dtype,
+            ),
+            requires_grad=False,
+        )
+        set_weight_attrs(
+            scales,
+            {
+                "output_dim": 0,
+                "packed_dim": 0,
+                "pack_factor": self.quant_config.out_group_size
+            },
+        )
+
+        return {
+            "codes": codes,
+            "codebooks": codebooks,
+            "scales": scales,
+        }
+
+    def apply_weights(
+        self,
+        weights: Dict[str, Any],
+        x: torch.Tensor,
+        bias: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        codebooks = weights["codebooks"]
+        codes = weights["codes"]
+        scales = weights["scales"]
+
+        shard_dim = getattr(codebooks, "shard_dim", None)
+        if shard_dim is not None:
```

Comment:
I need to understand more about how the codebooks work (let's sync offline about this), but I am curious to see if there is a way to avoid having to run this for loop

The other activation kernels have groups that are independent of eachother, which enable us to treat the weights and scales for the 3 logical matrices as one logical matrix rather than having to iterate over each separately

Do the codebooks not work this way?